### ISSUE TITLE
Certificate breaks decode_asn1:_parse_asn1_time

### ISSUE THREAD

**Issue by jcrowgey** [pyca/cryptography#4158]
Certificate found in the wild seem to break cryptography's date parser.  The underlying openssl tool handles the same data without crashing:

```
$ echo "MIICizCCAfQCCQD60p0z19VwkjANBgkqhkiG9w0BAQUFADCBhzELMAkGA1UEBhMCTkwxEzARBgNVBAgTClNvbWUtU3RhdGUxDzANBgNVBAcTBkFsbWVyZTEOMAwGA1UEChMFTUJJY3QxFDASBgNVBAsTC2RldmVsb3BtZW50MQ4wDAYDVQQDEwVNQkljdDEcMBoGCSqGSIb3DQEJARYNaW5mb0BtYmljdC5ubDAiFw0xNTA2MjAxOTQxMTNaGBEyNzU4MDYwMzAyMTk0MTEzWjCBhzELMAkGA1UEBhMCTkwxEzARBgNVBAgTClNvbWUtU3RhdGUxDzANBgNVBAcTBkFsbWVyZTEOMAwGA1UEChMFTUJJY3QxFDASBgNVBAsTC2RldmVsb3BtZW50MQ4wDAYDVQQDEwVNQkljdDEcMBoGCSqGSIb3DQEJARYNaW5mb0BtYmljdC5ubDCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA7XfoqQkpO2AJ/i+4vO0qzoKdRRo8YIlLgmwiry4r/M854yzKrujjdhLc3AjqqUy1w+FWR+THBdSOWXNPHtDuHsjZAE1A9bWpoZFBAIOTsMvQp8W3C1DK77umVkf1+i924aUPCeJHKWQJ9nIjM8+uvEvrBfC1Kz8b7hHD5fdXWbcCAwEAATANBgkqhkiG9w0BAQUFAAOBgQAONDYlp/cTUaNBBm7OYtXJzk1avUE4/l5PTKZBqDrZVkC55zUT2D4kLHjARXD08xLfsppEbPovo637lpSqsUI0eLHndVb5S/97TZaSpdY3SNMbDQXwmGdsYnjRGDeidGWcg3sCLifqFakWSuMI5Il+9S9nqlUBJYwF5CgCyv6H1g==" | base64 -d | openssl x509 -text -inform DER
Certificate:
    Data:
        Version: 1 (0x0)
        Serial Number:
            fa:d2:9d:33:d7:d5:70:92
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: C = NL, ST = Some-State, L = Almere, O = MBIct, OU = development, CN = MBIct, emailAddress = info@mbict.nl
        Validity
            Not Before: Jun 20 19:41:13 2015 GMT
            Not After : Jun  3 02:19:41 2758 GMT
        Subject: C = NL, ST = Some-State, L = Almere, O = MBIct, OU = development, CN = MBIct, emailAddress = info@mbict.nl
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (1024 bit)
                Modulus:
                    00:ed:77:e8:a9:09:29:3b:60:09:fe:2f:b8:bc:ed:
                    2a:ce:82:9d:45:1a:3c:60:89:4b:82:6c:22:af:2e:
                    2b:fc:cf:39:e3:2c:ca:ae:e8:e3:76:12:dc:dc:08:
                    ea:a9:4c:b5:c3:e1:56:47:e4:c7:05:d4:8e:59:73:
                    4f:1e:d0:ee:1e:c8:d9:00:4d:40:f5:b5:a9:a1:91:
                    41:00:83:93:b0:cb:d0:a7:c5:b7:0b:50:ca:ef:bb:
                    a6:56:47:f5:fa:2f:76:e1:a5:0f:09:e2:47:29:64:
                    09:f6:72:23:33:cf:ae:bc:4b:eb:05:f0:b5:2b:3f:
                    1b:ee:11:c3:e5:f7:57:59:b7
                Exponent: 65537 (0x10001)
    Signature Algorithm: sha1WithRSAEncryption
         0e:34:36:25:a7:f7:13:51:a3:41:06:6e:ce:62:d5:c9:ce:4d:
         5a:bd:41:38:fe:5e:4f:4c:a6:41:a8:3a:d9:56:40:b9:e7:35:
         13:d8:3e:24:2c:78:c0:45:70:f4:f3:12:df:b2:9a:44:6c:fa:
         2f:a3:ad:fb:96:94:aa:b1:42:34:78:b1:e7:75:56:f9:4b:ff:
         7b:4d:96:92:a5:d6:37:48:d3:1b:0d:05:f0:98:67:6c:62:78:
         d1:18:37:a2:74:65:9c:83:7b:02:2e:27:ea:15:a9:16:4a:e3:
         08:e4:89:7e:f5:2f:67:aa:55:01:25:8c:05:e4:28:02:ca:fe:
         87:d6
-----BEGIN CERTIFICATE-----
MIICizCCAfQCCQD60p0z19VwkjANBgkqhkiG9w0BAQUFADCBhzELMAkGA1UEBhMC
TkwxEzARBgNVBAgTClNvbWUtU3RhdGUxDzANBgNVBAcTBkFsbWVyZTEOMAwGA1UE
ChMFTUJJY3QxFDASBgNVBAsTC2RldmVsb3BtZW50MQ4wDAYDVQQDEwVNQkljdDEc
MBoGCSqGSIb3DQEJARYNaW5mb0BtYmljdC5ubDAiFw0xNTA2MjAxOTQxMTNaGBEy
NzU4MDYwMzAyMTk0MTEzWjCBhzELMAkGA1UEBhMCTkwxEzARBgNVBAgTClNvbWUt
U3RhdGUxDzANBgNVBAcTBkFsbWVyZTEOMAwGA1UEChMFTUJJY3QxFDASBgNVBAsT
C2RldmVsb3BtZW50MQ4wDAYDVQQDEwVNQkljdDEcMBoGCSqGSIb3DQEJARYNaW5m
b0BtYmljdC5ubDCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA7XfoqQkpO2AJ
/i+4vO0qzoKdRRo8YIlLgmwiry4r/M854yzKrujjdhLc3AjqqUy1w+FWR+THBdSO
WXNPHtDuHsjZAE1A9bWpoZFBAIOTsMvQp8W3C1DK77umVkf1+i924aUPCeJHKWQJ
9nIjM8+uvEvrBfC1Kz8b7hHD5fdXWbcCAwEAATANBgkqhkiG9w0BAQUFAAOBgQAO
NDYlp/cTUaNBBm7OYtXJzk1avUE4/l5PTKZBqDrZVkC55zUT2D4kLHjARXD08xLf
sppEbPovo637lpSqsUI0eLHndVb5S/97TZaSpdY3SNMbDQXwmGdsYnjRGDeidGWc
g3sCLifqFakWSuMI5Il+9S9nqlUBJYwF5CgCyv6H1g==
-----END CERTIFICATE-----
```

Steps to reproduce:

1) create a virtualenv and pip install cryptography

```
$ virtualenv venv
Running virtualenv with interpreter /usr/bin/python2
New python executable in /home/j/src/cryptotest/venv/bin/python2
Also creating executable in /home/j/src/cryptotest/venv/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
$ . venv/bin/activate
(venv)$ pip install cryptography
Collecting cryptography
[snip]
```

2) try to pull the end validity date out of the certificate
 
```
(venv) $ python
Python 2.7.13 (default, Nov 24 2017, 17:33:09) 
[GCC 6.3.0 20170516] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import cryptography
>>> import base64
>>> text = 'MIICizCCAfQCCQD60p0z19VwkjANBgkqhkiG9w0BAQUFADCBhzELMAkGA1UEBhMCTkwxEzARBgNVBAgTClNvbWUtU3RhdGUxDzANBgNVBAcTBkFsbWVyZTEOMAwGA1UEChMFTUJJY3QxFDASBgNVBAsTC2RldmVsb3BtZW50MQ4wDAYDVQQDEwVNQkljdDEcMBoGCSqGSIb3DQEJARYNaW5mb0BtYmljdC5ubDAiFw0xNTA2MjAxOTQxMTNaGBEyNzU4MDYwMzAyMTk0MTEzWjCBhzELMAkGA1UEBhMCTkwxEzARBgNVBAgTClNvbWUtU3RhdGUxDzANBgNVBAcTBkFsbWVyZTEOMAwGA1UEChMFTUJJY3QxFDASBgNVBAsTC2RldmVsb3BtZW50MQ4wDAYDVQQDEwVNQkljdDEcMBoGCSqGSIb3DQEJARYNaW5mb0BtYmljdC5ubDCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA7XfoqQkpO2AJ/i+4vO0qzoKdRRo8YIlLgmwiry4r/M854yzKrujjdhLc3AjqqUy1w+FWR+THBdSOWXNPHtDuHsjZAE1A9bWpoZFBAIOTsMvQp8W3C1DK77umVkf1+i924aUPCeJHKWQJ9nIjM8+uvEvrBfC1Kz8b7hHD5fdXWbcCAwEAATANBgkqhkiG9w0BAQUFAAOBgQAONDYlp/cTUaNBBm7OYtXJzk1avUE4/l5PTKZBqDrZVkC55zUT2D4kLHjARXD08xLfsppEbPovo637lpSqsUI0eLHndVb5S/97TZaSpdY3SNMbDQXwmGdsYnjRGDeidGWcg3sCLifqFakWSuMI5Il+9S9nqlUBJYwF5CgCyv6H1g=='
>>> from cryptography import x509
>>> cert = cryptography.x509.load_der_x509_certificate(base64.b64decode(text), cryptography.hazmat.backends.default_backend())
>>> cert.not_valid_after
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/j/src/cryptotest/venv/local/lib/python2.7/site-packages/cryptography/hazmat/backends/openssl/x509.py", line 96, in not_valid_after
    return _parse_asn1_time(self._backend, asn1_time)
  File "/home/j/src/cryptotest/venv/local/lib/python2.7/site-packages/cryptography/hazmat/backends/openssl/decode_asn1.py", line 737, in _parse_asn1_time
    backend.openssl_assert(generalized_time != backend._ffi.NULL)
  File "/home/j/src/cryptotest/venv/local/lib/python2.7/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 105, in openssl_assert
    return binding._openssl_assert(self._lib, ok)
  File "/home/j/src/cryptotest/venv/local/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 76, in _openssl_assert
    errors_with_text
cryptography.exceptions.InternalError: Unknown OpenSSL error. This error is commonly encountered when another library is not cleaning up the OpenSSL error stack. If you are using cryptography with another library that uses OpenSSL try disabling it before reporting a bug. Otherwise please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([])
```

**Comment by jcrowgey** [comment]
Possibly related.  This certificate causes a ValueError:

```
-----BEGIN CERTIFICATE-----
MIIDfzCCAuigAwIBAgIJAIJqsyH6U4xqMA0GCSqGSIb3DQEBBQUAMIGQMQswCQYD
VQQGEwJjaDEQMA4GA1UEBxQHWsO8cmljaDEcMBoGA1UEChQTQS5ILiBNZXllciAm
IENpZSBBRzEoMCYGA1UEAxQfQS5ILiBNZXllciAmIENpZSBBRyBXZWJBZG1pbiBD
QTEnMCUGCSqGSIb3DQEJARYYYWRtaW5pc3RyYXRvckBhaG1leWVyLmNoMCIXDTA5
MDkwOTEzNTgyMFoXETM3MDEyNDEzNTgyNCswMDAwMFExCzAJBgNVBAYTAmNoMRAw
DgYDVQQHFAdaw7xyaWNoMRwwGgYDVQQKFBNBLkguIE1leWVyICYgQ2llIEFHMRIw
EAYDVQQDEwlhaG1oZnctMDEwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAKoT
u5uM1hk6II/a3DCm8gSeVUyZwyxHvNurS6rztZ/vaQhC7P5Bze/zIfenK7WW3uy8
Fz+Re25qfxA4KTRYI1FjBstqPjCFxYV+X40k5q1MVE6Rx2HYpMi8dvLVgjywmdDl
LjZ0h475Mxa3MC/oQ1Sco7ncV4Rgjm84H1zUj6hVAgMBAAGjggEZMIIBFTAdBgNV
HQ4EFgQUPCDwlBpaSR4aWCu4ZSGy/E0VWIIwgcUGA1UdIwSBvTCBuoAUj+bMC3xr
GCUDl+Sew0WKRbMtMxWhgZakgZMwgZAxCzAJBgNVBAYTAmNoMRAwDgYDVQQHFAda
w7xyaWNoMRwwGgYDVQQKFBNBLkguIE1leWVyICYgQ2llIEFHMSgwJgYDVQQDFB9B
LkguIE1leWVyICYgQ2llIEFHIFdlYkFkbWluIENBMScwJQYJKoZIhvcNAQkBFhhh
ZG1pbmlzdHJhdG9yQGFobWV5ZXIuY2iCCQCCarMh+lOMaTAUBgNVHREEDTALgglh
aG1oZnctMDEwCQYDVR0TBAIwADALBgNVHQ8EBAMCBeAwDQYJKoZIhvcNAQEFBQAD
gYEAk6mJFRJxGOtNXI1cFDKMeonPsMCp7cwuyV9QrO70bznPQ+MOLWfzV5jAJ3v2
rLCwhW203hnvPvLACDKX6LtcvxPnHCgy4ipvamw0IcE+hCeTT2RHa5dh8acymn4O
pXpgA96g6GFG76nM/Nlpy/hEOZO/cnF/UHwoX9v2sF72EJc=
-----END CERTIFICATE-----
```

Different error, this one raises a ValueError from datetime, but I can't help but notice that it's still coming from `_parse_asn1_time`

```
    na = certificate.not_valid_after
  File "/home/j/src/x509parser/venv/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/x509.py", line 96, in not_valid_after
    return _parse_asn1_time(self._backend, asn1_time)
  File "/home/j/src/x509parser/venv/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/decode_asn1.py", line 741, in _parse_asn1_time
    return _parse_asn1_generalized_time(backend, generalized_time)
  File "/home/j/src/x509parser/venv/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/decode_asn1.py", line 748, in _parse_asn1_generalized_time
    return datetime.datetime.strptime(time, "%Y%m%d%H%M%SZ")
  File "/usr/lib/python3.5/_strptime.py", line 510, in _strptime_datetime
    tt, fraction = _strptime(data_string, format)
  File "/usr/lib/python3.5/_strptime.py", line 343, in _strptime
    (data_string, format))
ValueError: time data '20370124135824+0000' does not match format '%Y%m%d%H%M%SZ'
```

**Comment by tdsmith** [comment]
I just found another example that gives an InternalError:

```
x509.load_pem_x509_certificate(b"""
-----BEGIN CERTIFICATE-----
MIIFdDCCBFygAwIBAgIQJ2buVutJ846r13Ci/ITeIjANBgkqhkiG9w0BAQwFADBv
MQswCQYDVQQGEwJTRTEUMBIGA1UEChMLQWRkVHJ1c3QgQUIxJjAkBgNVBAsTHUFk
ZFRydXN0IEV4dGVybmFsIFRUUCBOZXR3b3JrMSIwIAYDVQQDExlBZGRUcnVzdCBF
eHRlcm5hbCBDQSBSb290MB4XDTAwMDUzMDEwNDgzOFoXDTIwMDUzODEwNDg3OHow
gYUxCzAJBgNVBAYTAkdCMRswGQYDVQQIExJHcmVhdGVyIE1hbmNoZXN0ZXIxEDAO
BgNVBAcTB1NhbGZvcmQxGjAYBgNVBAoTEUNPTU9ETyBDQSBMaW1pdGVkMSswKQYD
VQQDEyJDT01PRE8gUlNBIENlcnRpZmljYXRpb24gQXV0aG9yaXR5MIICIjANBgkq
hkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAkehUktIKVrGsDSTdxc9EZ3SZKzejfSNw
AHG8U9/E+ioSj0t/EFa9n3Byt2F/yUsPF6c947AEYe7/EZfH9IY+Cvo+XPmT5jR6
2RRr55yzhaCCenavcZDX7P0N+pxs+t+wgvQUfvm+xKYvT3+Zf7X8Z0NyvQwA1onr
ayzT7Y+YHBSrfuXjbvzYqOSSJNpDa2K4Vf3qwbxstovzDo2a5JtsaZn4eEgwRdWt
4Q08RWD8MpZRJ7xnw8outmvqRsfHIKCxH2XeSAi6pE6p8oNGN4Tr6MyBSENnTnIq
m1y9TBsoilwie7SrmNnu4FGDwwlGTm0+mfqVF9p8M1dBPI1R7Qu2XK8sYxrfV8g/
vOldxJuvRZnio1oktLqpVj3Pb6r/SVi+8Kj/9Lit6Tf7urj0Czr56ENCHonYhMsT
8dm74YlguIwoVqwUHZwK53Hrzw7dPamWoUi9PPevtQ0iTMARgexWO/bTouJbt7IE
IlKVgJNp6I5MZfGRAy1wdALqi2cVKWlSArvX31BqVUa/oKMoYX9w0MOiqiwhqkfO
KJwGRXa/ghgntNWutMtQ5mv0TIZxMOmm3xaG4Nj/QN370EKIf6MzOi5cHkERgWPO
GHFrK+ymircxXDpqR+DDeVnWIBqv8mqYqnK8V0rSS527EPywTEHl7R09XiidnMy/
s1Hap0flhFMCAwEAAaOB9DCB8TAfBgNVHSMEGDAWgBStvZh6NLQm9/rEJlTvA73g
JMtUGjAdBgNVHQ4EFgQUu69+Aj36pvE8hI6t7jiY7NkyMtQwDgYDVR0PAQH/BAQD
AgGGMA8GA1UdEwEB/wQFMAMBAf8wEQYDVR0gBAowCDAGBgRVHSAAMEQGA1UdHwQ9
MDswOaA3oDWGM2h0dHA6Ly9jcmwudXNlcnRydXN0LmNvbS9BZGRUcnVzdEV4dGVy
bmFsQ0FSb290LmNybDA1BggrBgEFBQcBAQQpMCcwJQYIKwYBBQUHMAGGGWh0dHA6
Ly9vY3NwLnVzZXJ0cnVzdC5jb20wDQYJKoZIhvcNAQEMBQADggEBAGS/g/FfmoXQ
zbihKVcN6Fr30ek+8nYEbvFScLsePP9NDXRqzIGCJdPDoCpdTPW6i6FtxFQJdcfj
Jw5dhHk3QBN39bSsHNA7qxcS1u80GH4r6XnTq1dFDK8o+tDb5VCViLvfhVdpfZLY
Uspzgb8c8+a4bmYRBbMelC1/kZWSWfFMzqORcUx8Rww7Cxn2obFshj5cqsQugsv5
B5a6SE2Q8pTIqXOi6wZ7I53eovNNVZ96YUWYGGjHXkBrI/V5eu+MtWuLt29G9Hvx
PUsE2JOAWVrgQSQdso8VYFhH2+9uRv0V9dlfmrPb2LjkQLPNlzmuhbsdjrzch5vR
pu/xO28QOG8=
-----END CERTIFICATE-----
""", backend).not_valid_after
```
<details>
<summary>Traceback</summary>

```
~/.virtualenvs/ct/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/x509.py in not_valid_after(self)
     94     def not_valid_after(self):
     95         asn1_time = self._backend._lib.X509_get_notAfter(self._x509)
---> 96         return _parse_asn1_time(self._backend, asn1_time)
     97 
     98     @property

~/.virtualenvs/ct/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/decode_asn1.py in _parse_asn1_time(backend, asn1_time)
    735         asn1_time, backend._ffi.NULL
    736     )
--> 737     backend.openssl_assert(generalized_time != backend._ffi.NULL)
    738     generalized_time = backend._ffi.gc(
    739         generalized_time, backend._lib.ASN1_GENERALIZEDTIME_free

~/.virtualenvs/ct/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/backend.py in openssl_assert(self, ok)
    104 
    105     def openssl_assert(self, ok):
--> 106         return binding._openssl_assert(self._lib, ok)
    107 
    108     def activate_builtin_random(self):

~/.virtualenvs/ct/lib/python3.6/site-packages/cryptography/hazmat/bindings/openssl/binding.py in _openssl_assert(lib, ok)
     73             "issues with information on how to reproduce "
     74             "this. ({0!r})".format(errors_with_text),
---> 75             errors_with_text
     76         )
     77 

InternalError: Unknown OpenSSL error. This error is commonly encountered when another library is not cleaning up the OpenSSL error stack. If you are using cryptography with another library that uses OpenSSL try disabling it before reporting a bug. Otherwise please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([])
```
</details>

**Comment by alex** [comment]
That's not a real date

**Comment by reaperhulk** [comment]
May 38 😆 . Gotta love string-based time.

**Comment by jcrowgey** [comment]
I found another certificate which seems to cause the same error reported in the first post:

```
>>> backend = cryptography.hazmat.backends.default_backend()
>>> data = base64.b64decode(b"MIIC9DCCAdygAwIBAgIJAM/N3nz0v2b9MA0GCSqGSIb3DQEBBQUAMC8xCzAJBgNVBAYTAkdCMSAwHgYDVQQDExdmY2tQQTIwMC5xdWlja2Vuam95LmNvbTAgFw0xMTAzMjEwOTI1NTJaFw8xOTAyMDcwMTAyNTczNlowgZUxCzAJBgNVBAYTAlVTMRcwFQYDVQQIEw5EZWZhdWx0IFJlZ2lvbjEZMBcGA1UEBxMQRGVmYXVsdCBMb2NhbGl0eTEdMBsGA1UEChMURGVmYXVsdCBPcmdhbml6YXRpb24xFTATBgNVBAsTDERlZmF1bHQgVW5pdDEcMBoGA1UEAxMTRGVmYXVsdCBDb21tb24gTmFtZTCBnjANBgkqhkiG9w0BAQEFAAOBjAAwgYgCgYDHWlVCGI2r1fKY7YawkjNaeNZ77cCkQsfWw4AdgmOdHtt9xJdkbCHGx7Yzh7BRu6QkvjQKwgecdb2GZ7zRAF8FXPFa5fJbozQSJVqMMKCnTtm6PjgA4jgH3m7ov/JOcFsXVmWtQHFAnE6r2s6QE59LbryTb0cvmk9Pq63WxLenAQIDAQABoy8wLTAdBgNVHQ4EFgQU+upJrZrv9MLDSfoVdo0vMdc5rzgwDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQUFAAOCAQEAqjD19zQ36dVjCmPkSv/cjGfhiRtKV3Q3UVRFnDX0Lk5BB+GqeVgVfJ8oQForVlR1rZIAbnl0St+/xtUYRhBss5jRIKHkoylc5k1T7jU8fFzTxjbenPlrt5V5+IH+UnLe5ozpoTsp69TgpQ3ImYDuxun73+OVSE6SLKBBoy0BCrNG1ezg6P6idqR3rCS3cQFfmP9nJHcOgBJ71bvAbzfc+pmlliOPG1SVOGYnxLJjduLYELHUN1hgSH+bcQtQMmsxkixXq3ejYvWyir4W3CPflwaUfXBes18SjVN91T+cKBU5qlWMLRuYqmyCq9oiER2WSZH05FeHwhYyPjOuf/fIgA==")
>>> cert = cryptography.x509.load_der_x509_certificate(data, backend)
>>> cert.not_valid_after
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/j/src/x509parser/venv/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/x509.py", line 96, in not_valid_after
    return _parse_asn1_time(self._backend, asn1_time)
  File "/home/j/src/x509parser/venv/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/decode_asn1.py", line 737, in _parse_asn1_time
    backend.openssl_assert(generalized_time != backend._ffi.NULL)
  File "/home/j/src/x509parser/venv/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 105, in openssl_assert
    return binding._openssl_assert(self._lib, ok)
  File "/home/j/src/x509parser/venv/lib/python3.5/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 76, in _openssl_assert
    errors_with_text
cryptography.exceptions.InternalError: Unknown OpenSSL error. This error is commonly encountered when another library is not cleaning up the OpenSSL error stack. If you are using cryptography with another library that uses OpenSSL try disabling it before reporting a bug. Otherwise please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([])
```

openssl x509 seems ok with this data:

```
$ echo -n "MIIC9DCCAdygAwIBAgIJAM/N3nz0v2b9MA0GCSqGSIb3DQEBBQUAMC8xCzAJBgNVBAYTAkdCMSAwHgYDVQQDExdmY2tQQTIwMC5xdWlja2Vuam95LmNvbTAgFw0xMTAzMjEwOTI1NTJaFw8xOTAyMDcwMTAyNTczNlowgZUxCzAJBgNVBAYTAlVTMRcwFQYDVQQIEw5EZWZhdWx0IFJlZ2lvbjEZMBcGA1UEBxMQRGVmYXVsdCBMb2NhbGl0eTEdMBsGA1UEChMURGVmYXVsdCBPcmdhbml6YXRpb24xFTATBgNVBAsTDERlZmF1bHQgVW5pdDEcMBoGA1UEAxMTRGVmYXVsdCBDb21tb24gTmFtZTCBnjANBgkqhkiG9w0BAQEFAAOBjAAwgYgCgYDHWlVCGI2r1fKY7YawkjNaeNZ77cCkQsfWw4AdgmOdHtt9xJdkbCHGx7Yzh7BRu6QkvjQKwgecdb2GZ7zRAF8FXPFa5fJbozQSJVqMMKCnTtm6PjgA4jgH3m7ov/JOcFsXVmWtQHFAnE6r2s6QE59LbryTb0cvmk9Pq63WxLenAQIDAQABoy8wLTAdBgNVHQ4EFgQU+upJrZrv9MLDSfoVdo0vMdc5rzgwDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQUFAAOCAQEAqjD19zQ36dVjCmPkSv/cjGfhiRtKV3Q3UVRFnDX0Lk5BB+GqeVgVfJ8oQForVlR1rZIAbnl0St+/xtUYRhBss5jRIKHkoylc5k1T7jU8fFzTxjbenPlrt5V5+IH+UnLe5ozpoTsp69TgpQ3ImYDuxun73+OVSE6SLKBBoy0BCrNG1ezg6P6idqR3rCS3cQFfmP9nJHcOgBJ71bvAbzfc+pmlliOPG1SVOGYnxLJjduLYELHUN1hgSH+bcQtQMmsxkixXq3ejYvWyir4W3CPflwaUfXBes18SjVN91T+cKBU5qlWMLRuYqmyCq9oiER2WSZH05FeHwhYyPjOuf/fIgA==" | base64 -d | openssl x509 -inform DER -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            cf:cd:de:7c:f4:bf:66:fd
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: C = GB, CN = fckPA200.quickenjoy.com
        Validity
            Not Before: Mar 21 09:25:52 2011 GMT
            Not After : Feb  7 01:02:57 2019 GMT
        Subject: C = US, ST = Default Region, L = Default Locality, O = Default Organization, OU = Default Unit, CN = Default Common Name
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (1024 bit)
                Modulus:
                    00:c7:5a:55:42:18:8d:ab:d5:f2:98:ed:86:b0:92:
                    33:5a:78:d6:7b:ed:c0:a4:42:c7:d6:c3:80:1d:82:
                    63:9d:1e:db:7d:c4:97:64:6c:21:c6:c7:b6:33:87:
                    b0:51:bb:a4:24:be:34:0a:c2:07:9c:75:bd:86:67:
                    bc:d1:00:5f:05:5c:f1:5a:e5:f2:5b:a3:34:12:25:
                    5a:8c:30:a0:a7:4e:d9:ba:3e:38:00:e2:38:07:de:
                    6e:e8:bf:f2:4e:70:5b:17:56:65:ad:40:71:40:9c:
                    4e:ab:da:ce:90:13:9f:4b:6e:bc:93:6f:47:2f:9a:
                    4f:4f:ab:ad:d6:c4:b7:a7:01
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                FA:EA:49:AD:9A:EF:F4:C2:C3:49:FA:15:76:8D:2F:31:D7:39:AF:38
            X509v3 Basic Constraints: 
                CA:TRUE
    Signature Algorithm: sha1WithRSAEncryption
         aa:30:f5:f7:34:37:e9:d5:63:0a:63:e4:4a:ff:dc:8c:67:e1:
         89:1b:4a:57:74:37:51:54:45:9c:35:f4:2e:4e:41:07:e1:aa:
         79:58:15:7c:9f:28:40:5a:2b:56:54:75:ad:92:00:6e:79:74:
         4a:df:bf:c6:d5:18:46:10:6c:b3:98:d1:20:a1:e4:a3:29:5c:
         e6:4d:53:ee:35:3c:7c:5c:d3:c6:36:de:9c:f9:6b:b7:95:79:
         f8:81:fe:52:72:de:e6:8c:e9:a1:3b:29:eb:d4:e0:a5:0d:c8:
         99:80:ee:c6:e9:fb:df:e3:95:48:4e:92:2c:a0:41:a3:2d:01:
         0a:b3:46:d5:ec:e0:e8:fe:a2:76:a4:77:ac:24:b7:71:01:5f:
         98:ff:67:24:77:0e:80:12:7b:d5:bb:c0:6f:37:dc:fa:99:a5:
         96:23:8f:1b:54:95:38:66:27:c4:b2:63:76:e2:d8:10:b1:d4:
         37:58:60:48:7f:9b:71:0b:50:32:6b:31:92:2c:57:ab:77:a3:
         62:f5:b2:8a:be:16:dc:23:df:97:06:94:7d:70:5e:b3:5f:12:
         8d:53:7d:d5:3f:9c:28:15:39:aa:55:8c:2d:1b:98:aa:6c:82:
         ab:da:22:11:1d:96:49:91:f4:e4:57:87:c2:16:32:3e:33:ae:
         7f:f7:c8:80
```

**Comment by reaperhulk** [comment]
The encoding on the not after time is a UTC time with value "19020701025736Z". That last "36" is invalid because the only allowed forms for UTCTime are:

```
YYMMDDhhmmZ
YYMMDDhhmm+hh'mm'
YYMMDDhhmm-hh'mm'
YYMMDDhhmmssZ
YYMMDDhhmmss+hh'mm'
YYMMDDhhmmss-hh'mm'
```

In contrast to the previous issue (where there's literally no way to parse a datetime from nonsense like "May 38") we could hypothetically work around this. However, I'm reluctant to do so since this is not allowable by spec.

**Comment by alex** [comment]
We should cleanly raise a ValueError, instead of an InternalError, but that's it.

**Comment by jcrowgey** [comment]
FWIW: I believe this is actually not fixed until #4180, which isn't yet merged.  #4179 adds the relevant tests vector.
